cpp-client/Dockerfile

```diff
@@ -26,9 +26,41 @@ RUN set -eux; \
 #
 # Download and build dependent libraries.
 #
+# Be mindful about the order below:
+# 1. Some libraries have cross dependencies (eg, arrow depends on grpc).
+# 2. If something fails, the docker cache at that point gets invalidated;
+#    so it makes sense to put more brittle stuff later on.
```

Comment:
We can use docker multi-stage builds to be more explicitly about the ordering if we want; and then re-join the built assets at the end.

Reply:
I started drafting this but looked like too much complexity; essentially I named the base os image using `FROM ... AS osbase` at the top, and then before each RUN command I did, eg,

```
FROM osbase as protobuf
RUN --mount=type=bind,source=build-dependencies.sh,target=/build-dependencies.sh \
    set -eux; mkdir -p /cpp-client/deps; cd /cpp-client/deps; \
    bash -c "/build-dependencies.sh --clean protobuf"
```

and then when dependencies were relevant, eg:
```
FROM osbase as grpc
COPY --from=protobuf /cpp-client/deps/local/protobuf /cpp-client/deps/local/protobuf/
```

and then at the end a bunch of copies.
But it ends up being pretty unwiedly I think, the sequence of RUNS alone in a particular order seems simpler and not different enough to justify all the additional fandango IMHO.

Reply:
That's fine - we don't need to do multi-stage. It is a lot of extra work for not much gain atm IMO.